### ISSUE TITLE
TEI to JSON conversion gives strange behavior with <seg/>

### ISSUE THREAD

**Issue by jjmccollum** [d-flood/Tendon#1]
While working on collating the IGNTP's TEI XML transcriptions in Ephesians, I noticed some unusual behavior in how Criticus converted the transcription for GA 056. This witness begins with some lectionary notes at the start of Eph 1:1 (encoded as `<seg/>` elements containing `<fw/>` children):

```xml
<div type="chapter" n="B10K1">
  <ab n="B10K1V1">
    <seg type="margin" subtype="lineright" n="@P246rC1L-056">
      <fw type="lectTitle" rend="left">
        <w>τη</w>
        <w><abbr type="num"><hi rend="overline">ε</hi></abbr></w>
        <w>τ<ex>ης</ex></w>
        <w><abbr type="num"><hi rend="overline">ιστ</hi></abbr></w>
        <w>εβδ<ex>ομαδος</ex></w>
      </fw>
    </seg>
    <w>παυλος</w><w>αποστολος</w>
    ...
  </ab>
</div>
```

These elements appear again in Eph 1:7, 1:9 (twice), 1:16, and again in 1:17, just before the text εν επιγνωσει αυτου:

```xml
<ab n="B10K1V17">
  <w>ινα</w>
  <w>ο</w>
  <w><abbr type="nomSac"><hi rend="overline">θς</hi></abbr></w>
  <w>του</w>
  <w><abbr type="nomSac"><hi rend="overline">κυ</hi></abbr></w>
  <w>ημων</w>
  <w><abbr type="nomSac"><hi rend="overline">ιυ</hi></abbr></w>
  <w><abbr type="nomSac"><hi rend="overline">χυ</hi></abbr></w>
  <w>ο</w>
  <w><abbr type="nomSac"><hi rend="overline">πηρ</hi></abbr></w>
  <w>της</w>
  <w>δοξης</w>
  <w>δωη</w>
  <w>υμιν</w>
  <w><abbr type="nomSac"><hi rend="overline">πνα</hi></abbr></w>
  <w>σο<lb n="P248rC1L-056" break="no"/>φιας</w>
  <w><ex>και</ex></w>
  <w>αποκαλυψεως</w>
  <seg type="line" subtype="above" n="@P248rC1L-056">
    <fw type="lectTitle">
      <w>τελ<ex>ος</ex></w>
    </fw>
  </seg>
  <seg type="margin" subtype="lineright" n="@P248rC1L-056">
    <fw type="lectTitle" rend="center">
      <w>τελο<ex>ς</ex></w>
    </fw>
  </seg>
  <w>εν</w>
  <w>επιγνωσει</w>
  <w>αυτου</w>
  <pc>·</pc>
</ab>
```

But when I convert the XML file to JSON using Criticus, I get the following content for Eph 1:1:

```json
{
    "id": "056_B10K1V1",
    "transcription": "056",
    "transcription_siglum": "056",
    "siglum": "056",
    "context": "B10K1V1",
    "n": "B10K1V1",
    "plain_text": "εν επιγνωσει αυτου",
    "witnesses": [
        {
            "id": "056",
            "tokens": [
                {
                    "index": "2",
                    "siglum": "056",
                    "reading": "056",
                    "original": "εν",
                    "rule_match": [
                        "εν"
                    ],
                    "t": "εν"
                },
                {
                    "index": "4",
                    "siglum": "056",
                    "reading": "056",
                    "original": "επιγνωσει",
                    "rule_match": [
                        "επιγνωσει"
                    ],
                    "t": "επιγνωσει"
                },
                {
                    "index": "6",
                    "siglum": "056",
                    "reading": "056",
                    "original": "αυτου",
                    "rule_match": [
                        "αυτου"
                    ],
                    "t": "αυτου"
                }
            ]
        }
    ]
}
```

So the parser seems to be skipping from 1:1 to 1:17 for some reason. I suspect it has something to do with the `<seg/>` and `<fw/>` elements; does the `from_tei.py` script need to know to ignore them?

If you need the TEI XML transcription of GA 056 to test things on your end, just let me know, and I'll e-mail it to you. Thanks!

**Comment by jjmccollum** [comment]
It also looks like the TEI XML transcription of GA 056 also uses `<seg/>` elements (without `<fw/>` children) within `<rdg/>` elements to encode corrections written over the line or in other locations. This is probably only an issue one would encounter in IGNTP transcriptions of majuscules (since they prefer to note many more details of this nature in majuscule transcriptions), but I thought it might be worth mentioning, as it is related to this issue.

**Comment by jjmccollum** [comment]
Modifying `from_tei.py` to include the following lines would be a start:

```python
text = re.sub(r'<seg[^<>]*>', '', text)
text = re.sub(r'</seg>', '', text)
```

This would at least take care of the instances in which `<seg/>` is used to denote the placement of a `<rdg/>`. The `<fw/>` elements should then be (correctly) skipped in the `get_all_words_in_verse` method.

**Comment by jjmccollum** [comment]
Actually, it might need to be a bit more involved; a look at the IGNTP transcription of GA 012 made it clear to me that `<seg/>` elements are also used by themselves in-text (and not just in `<rdg/>` elements) to indicate marginal notes. So this might be better:

```python
text = re.sub(r'(<rdg[^<>]*>)<seg[^<>]*>', r'\1', text)
text = re.sub(r'</seg></rdg>', '</rdg>', text)
```

**Comment by d-flood** [comment]
Okay, so I *think* most of this issue was resolved when I implemented your suggested fix for [the problem handling `<ex/>` elems](#2).

It looks to me like Eph 1:1, 7, 9, and 17 are correctly converted now.

I also looked into the use of `<seg/>` elements in 012. All the problem-causers seem to be in relation to corrections within `<app/>` elements. I've addressed this by specifically handling `<seg/>` elements when a correction is detected and sent to the `handle_app` function.

An aside: While double-checking the output, I realized that the IGNTP transcription viewer on epistulae.org places the marginal readings in the text with no indication that they do not belong to the main text. Curious.

Eph 1:1 for 056 now comes out like this:
```json
{
    "id": "056_B10K1V1",
    "transcription": "056",
    "transcription_siglum": "056",
    "siglum": "056",
    "context": "B10K1V1",
    "n": "B10K1V1",
    "plain_text": "παυλος αποστολος ιυ χυ δια θεληματος θυ τοις αγιοις τοις ουσιν εν εφεσω και πιστοις εν χω ιυ",
    "witnesses": [
        {
            "id": "056",
            "tokens": [
                {
                    "index": "2",
                    "siglum": "056",
                    "reading": "056",
                    "original": "παυλος",
                    "rule_match": [
                        "παυλος"
                    ],
                    "t": "παυλος"
                },
                {
                    "index": "4",
                    "siglum": "056",
                    "reading": "056",
                    "original": "αποστολος",
                    "rule_match": [
                        "αποστολος"
                    ],
                    "t": "αποστολος"
                },
                {
                    "index": "6",
                    "siglum": "056",
                    "reading": "056",
                    "original": "ιυ",
                    "rule_match": [
                        "ιυ"
                    ],
                    "t": "ιυ"
                },
                {
                    "index": "8",
                    "siglum": "056",
                    "reading": "056",
                    "original": "χυ",
                    "rule_match": [
                        "χυ"
                    ],
                    "t": "χυ"
                },
                {
                    "index": "10",
                    "siglum": "056",
                    "reading": "056",
                    "original": "δια",
                    "rule_match": [
                        "δια"
                    ],
                    "t": "δια"
                },
                {
                    "index": "12",
                    "siglum": "056",
                    "reading": "056",
                    "original": "θεληματος",
                    "rule_match": [
                        "θεληματος"
                    ],
                    "t": "θεληματος"
                },
                {
                    "index": "14",
                    "siglum": "056",
                    "reading": "056",
                    "original": "θυ",
                    "rule_match": [
                        "θυ"
                    ],
                    "t": "θυ"
                },
                {
                    "index": "16",
                    "siglum": "056",
                    "reading": "056",
                    "original": "τοις",
                    "rule_match": [
                        "τοις"
                    ],
                    "t": "τοις"
                },
                {
                    "index": "18",
                    "siglum": "056",
                    "reading": "056",
                    "original": "αγιοις",
                    "rule_match": [
                        "αγιοις"
                    ],
                    "t": "αγιοις"
                },
                {
                    "index": "20",
                    "siglum": "056",
                    "reading": "056",
                    "original": "τοις",
                    "rule_match": [
                        "τοις"
                    ],
                    "t": "τοις"
                },
                {
                    "index": "22",
                    "siglum": "056",
                    "reading": "056",
                    "original": "ουσιν",
                    "rule_match": [
                        "ουσιν"
                    ],
                    "t": "ουσιν"
                },
                {
                    "index": "24",
                    "siglum": "056",
                    "reading": "056",
                    "original": "εν",
                    "rule_match": [
                        "εν"
                    ],
                    "t": "εν"
                },
                {
                    "index": "26",
                    "siglum": "056",
                    "reading": "056",
                    "original": "εφεσω",
                    "rule_match": [
                        "εφεσω"
                    ],
                    "t": "εφεσω"
                },
                {
                    "index": "28",
                    "siglum": "056",
                    "reading": "056",
                    "original": "και",
                    "rule_match": [
                        "και"
                    ],
                    "t": "και"
                },
                {
                    "index": "30",
                    "siglum": "056",
                    "reading": "056",
                    "original": "πιστοις",
                    "rule_match": [
                        "πιστοις"
                    ],
                    "t": "πιστοις"
                },
                {
                    "index": "32",
                    "siglum": "056",
                    "reading": "056",
                    "original": "εν",
                    "rule_match": [
                        "εν"
                    ],
                    "t": "εν"
                },
                {
                    "index": "34",
                    "siglum": "056",
                    "reading": "056",
                    "original": "χω",
                    "rule_match": [
                        "χω"
                    ],
                    "t": "χω"
                },
                {
                    "index": "36",
                    "siglum": "056",
                    "reading": "056",
                    "original": "ιυ",
                    "rule_match": [
                        "ιυ"
                    ],
                    "t": "ιυ"
                }
            ]
        }
    ]
}
```

**Comment by d-flood** [comment]
This fix (and others) have (hopefully) been addressed in the [v0.33.0 PYPI release](https://pypi.org/project/criticus/0.33.0/)
Don't hesitate to reopen or start a new related issue.
